File: frontend/src/features/firecrawl/schemas.ts
import { z } from "zod";

export const AccountStatusSchema = z.enum(["active", "rate_limited", "credit_exhausted", "paused", "invalid"]);
export const CredentialStatusSchema = z.enum(["active", "paused", "invalid"]);
export const JobEndpointSchema = z.enum(["crawl", "batch_scrape"]);
export const SyncEndpointSchema = z.enum(["scrape", "map", "search"]);

export const FirecrawlCredentialSchema = z.object({
  id: z.string(),
  name: z.string().nullable(),
  status: CredentialStatusSchema,
});

export const FirecrawlAccountSchema = z.object({
  id: z.string(),
  team_label: z.string(),
  plan_type: z.string(),
  status: AccountStatusSchema,
  monthly_budget_credits: z.number().int().nullable(),
  remaining_credits_live: z.number().int().nullable(),
  plan_credits_live: z.number().int().nullable(),
  rpm_limit: z.number().int().nullable(),
  max_concurrency: z.number().int().nullable(),
  cooldown_until: z.string().nullable(),
  credentials: z.array(FirecrawlCredentialSchema),
});

export const FirecrawlAccountsResponseSchema = z.object({
  accounts: z.array(FirecrawlAccountSchema),
});

export const FirecrawlAccountCreateRequestSchema = z.object({
  id: z.string().trim().min(1),
  team_label: z.string().trim().min(1),
  plan_type: z.string().trim().min(1),
  monthly_budget_credits: z.number().int().min(0).nullable(),
  remaining_credits_live: z.number().int().min(0).nullable(),
  plan_credits_live: z.number().int().min(0).nullable(),
  rpm_limit: z.number().int().min(1).nullable(),
  max_concurrency: z.number().int().min(1).nullable(),
});

export const FirecrawlAccountUpdateRequestSchema = z.object({
  status: AccountStatusSchema.optional(),
  monthly_budget_credits: z.number().int().min(0).nullable().optional(),
  remaining_credits_live: z.number().int().min(0).nullable().optional(),
  plan_credits_live: z.number().int().min(0).nullable().optional(),
  rpm_limit: z.number().int().min(1).nullable().optional(),
  max_concurrency: z.number().int().min(1).nullable().optional(),
  cooldown_until: z.string().nullable().optional(),
});

export const FirecrawlCredentialCreateRequestSchema = z.object({
  id: z.string().trim().min(1),
  name: z.string().trim().min(1).nullable(),
  api_key: z.string().min(1),
});

export const FirecrawlCredentialUpdateRequestSchema = z.object({
  status: CredentialStatusSchema,
});

export const FirecrawlJobSchema = z.object({
  id: z.number().int(),
  account_id: z.string().nullable(),
  credential_id: z.string().nullable(),
  endpoint: JobEndpointSchema,
  upstream_job_id: z.string().nullable(),
  status: z.string(),
  estimated_credits_reserved: z.number().int().nullable(),
  credits_used_final: z.number().int().nullable(),
  created_at: z.string(),
  completed_at: z.string().nullable(),
  last_polled_at: z.string().nullable(),
});

export const FirecrawlJobsResponseSchema = z.object({
  jobs: z.array(FirecrawlJobSchema),
});

export const FirecrawlRequestLogSchema = z.object({
  id: z.number().int(),
  account_id: z.string().nullable(),
  credential_id: z.string().nullable(),
  endpoint: SyncEndpointSchema,
  upstream_job_id: z.string().nullable(),
  status: z.string(),
  upstream_status_code: z.number().int().nullable(),
  estimated_credits_pre: z.number().int().nullable(),
  credits_used_final: z.number().int().nullable(),
  latency_ms: z.number().int().nullable(),
  error_code: z.string().nullable(),
  error_message: z.string().nullable(),
  created_at: z.string(),
});

export const FirecrawlRequestLogsResponseSchema = z.object({
  logs: z.array(FirecrawlRequestLogSchema),
});

export const FirecrawlOverviewSchema = z.object({
  total_accounts: z.number().int(),
  active_accounts: z.number().int(),
  total_remaining_credits: z.number().int(),
  total_budget_credits: z.number().int(),
  accounts_by_status: z.object({
    active: z.number().int(),
    rate_limited: z.number().int(),
    credit_exhausted: z.number().int(),
    paused: z.number().int(),
  }),
  active_jobs: z.number().int(),
  recent_requests: z.object({
    total: z.number().int(),
    success: z.number().int(),
    error: z.number().int(),
  }),
  endpoint_breakdown: z.object({
    scrape: z.number().int(),
    map: z.number().int(),
    search: z.number().int(),
    crawl: z.number().int(),
    batch_scrape: z.number().int(),
  }),
});

export const FirecrawlRuntimeSettingsSchema = z.object({
  refresh_scheduler_enabled: z.boolean(),
  data_dir: z.string(),
  database_url_masked: z.string(),
  encryption_key_file: z.string(),
});

export type FirecrawlAccount = z.infer<typeof FirecrawlAccountSchema>;
export type FirecrawlAccountCreateRequest = z.infer<typeof FirecrawlAccountCreateRequestSchema>;
export type FirecrawlAccountUpdateRequest = z.infer<typeof FirecrawlAccountUpdateRequestSchema>;
export type FirecrawlCredential = z.infer<typeof FirecrawlCredentialSchema>;
export type FirecrawlCredentialCreateRequest = z.infer<typeof FirecrawlCredentialCreateRequestSchema>;
export type FirecrawlCredentialUpdateRequest = z.infer<typeof FirecrawlCredentialUpdateRequestSchema>;
export type FirecrawlJob = z.infer<typeof FirecrawlJobSchema>;
export type FirecrawlOverview = z.infer<typeof FirecrawlOverviewSchema>;
export type FirecrawlRequestLog = z.infer<typeof FirecrawlRequestLogSchema>;
export type FirecrawlRuntimeSettings = z.infer<typeof FirecrawlRuntimeSettingsSchema>;
export type JobEndpoint = z.infer<typeof JobEndpointSchema>;
export type SyncEndpoint = z.infer<typeof SyncEndpointSchema>;

File: app/modules/firecrawl/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_serializer

ACCOUNT_STATUS_PATTERN = r"^(active|rate_limited|credit_exhausted|paused|invalid)$"
CREDENTIAL_STATUS_PATTERN = r"^(active|paused|invalid)$"
JOB_ENDPOINT_PATTERN = r"^(crawl|batch_scrape)$"
REQUEST_LOG_ENDPOINT_PATTERN = r"^(scrape|map|search)$"


class FirecrawlAdminModel(BaseModel):
    model_config = ConfigDict(frozen=True)

    @field_serializer(
        "cooldown_until",
        "created_at",
        "completed_at",
        "last_polled_at",
        check_fields=False,
        when_used="json",
    )
    def serialize_datetime_as_utc(self, value: datetime | None) -> str | None:
        if isinstance(value, datetime):
            if value.tzinfo is None:
                return value.isoformat()
            return value.isoformat().replace("+00:00", "Z")
        return None


class FirecrawlCredentialResponse(FirecrawlAdminModel):
    id: str
    name: str | None
    status: str = Field(pattern=CREDENTIAL_STATUS_PATTERN)


class FirecrawlAccountResponse(FirecrawlAdminModel):
    id: str
    team_label: str
    plan_type: str
    status: str = Field(pattern=ACCOUNT_STATUS_PATTERN)
    monthly_budget_credits: int | None
    remaining_credits_live: int | None
    plan_credits_live: int | None
    rpm_limit: int | None
    max_concurrency: int | None
    cooldown_until: datetime | None
    credentials: list[FirecrawlCredentialResponse] = Field(default_factory=list)


class FirecrawlAccountsResponse(FirecrawlAdminModel):
    accounts: list[FirecrawlAccountResponse]


class FirecrawlJobResponse(FirecrawlAdminModel):
    id: int
    account_id: str | None
    credential_id: str | None
    endpoint: str = Field(pattern=JOB_ENDPOINT_PATTERN)
    upstream_job_id: str | None
    status: str
    estimated_credits_reserved: int | None
    credits_used_final: int | None
    created_at: datetime
    completed_at: datetime | None
    last_polled_at: datetime | None


class FirecrawlJobsResponse(FirecrawlAdminModel):
    jobs: list[FirecrawlJobResponse]


class FirecrawlRequestLogResponse(FirecrawlAdminModel):
    id: int
    account_id: str | None
    credential_id: str | None
    endpoint: str = Field(pattern=REQUEST_LOG_ENDPOINT_PATTERN)
    upstream_job_id: str | None
    status: str
    upstream_status_code: int | None
    estimated_credits_pre: int | None
    credits_used_final: int | None
    latency_ms: int | None
    error_code: str | None
    error_message: str | None
    created_at: datetime


class FirecrawlRequestLogsResponse(FirecrawlAdminModel):
    logs: list[FirecrawlRequestLogResponse]


class FirecrawlAccountsByStatusResponse(FirecrawlAdminModel):
    active: int
    rate_limited: int
    credit_exhausted: int
    paused: int


class FirecrawlRecentRequestsResponse(FirecrawlAdminModel):
    total: int
    success: int
    error: int


class FirecrawlEndpointBreakdownResponse(FirecrawlAdminModel):
    scrape: int
    map: int
    search: int
    crawl: int
    batch_scrape: int


class FirecrawlOverviewResponse(FirecrawlAdminModel):
    total_accounts: int
    active_accounts: int
    total_remaining_credits: int
    total_budget_credits: int
    accounts_by_status: FirecrawlAccountsByStatusResponse
    active_jobs: int
    recent_requests: FirecrawlRecentRequestsResponse
    endpoint_breakdown: FirecrawlEndpointBreakdownResponse


class FirecrawlAccountCreateRequest(FirecrawlAdminModel):
    id: str = Field(min_length=1)
    team_label: str = Field(min_length=1)
    plan_type: str = Field(default="unknown", min_length=1)
    monthly_budget_credits: int | None = Field(default=None, ge=0)
    remaining_credits_live: int | None = Field(default=None, ge=0)
    plan_credits_live: int | None = Field(default=None, ge=0)
    rpm_limit: int | None = Field(default=None, ge=1)
    max_concurrency: int | None = Field(default=None, ge=1)


class FirecrawlAccountUpdateRequest(FirecrawlAdminModel):
    status: str | None = Field(default=None, pattern=ACCOUNT_STATUS_PATTERN)
    monthly_budget_credits: int | None = Field(default=None, ge=0)
    remaining_credits_live: int | None = Field(default=None, ge=0)
    plan_credits_live: int | None = Field(default=None, ge=0)
    rpm_limit: int | None = Field(default=None, ge=1)
    max_concurrency: int | None = Field(default=None, ge=1)
    cooldown_until: datetime | None = None


class FirecrawlCredentialCreateRequest(FirecrawlAdminModel):
    id: str = Field(min_length=1)
    name: str | None = Field(default=None, min_length=1)
    api_key: str = Field(min_length=1)
    status: str = Field(default="active", pattern=CREDENTIAL_STATUS_PATTERN)


class FirecrawlCredentialUpdateRequest(FirecrawlAdminModel):
    status: str = Field(pattern=CREDENTIAL_STATUS_PATTERN)
